File: web-service/api-project/src/main/java/io/ehdev/conrad/service/api/service/user/UserTokenEndpoint.java
package io.ehdev.conrad.service.api.service.user;

import io.ehdev.conrad.model.permission.CreateTokenResponse;
import io.ehdev.conrad.model.permission.GetTokensResponse;
import io.ehdev.conrad.service.api.aop.annotation.LoggedInUserRequired;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import tech.crom.business.api.TokenManagementApi;
import tech.crom.model.token.GeneratedTokenDetails;
import tech.crom.model.token.TokenType;
import tech.crom.web.api.model.RequestDetails;

import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Controller
@RequestMapping("/api/v1/user/tokens")
public class UserTokenEndpoint {

    private final TokenManagementApi tokenManagementApi;

    @Autowired
    public UserTokenEndpoint(TokenManagementApi tokenManagementApi) {
        this.tokenManagementApi = tokenManagementApi;
    }

    @LoggedInUserRequired
    @RequestMapping(value = "/{tokenId}", method = RequestMethod.DELETE)
    public ResponseEntity deleteToken(RequestDetails requestDetails, @PathVariable("tokenId") String tokenId) {
        tokenManagementApi.invalidateToken(UUID.fromString(tokenId), TokenType.USER);
        return new ResponseEntity<>(HttpStatus.OK);
    }

    @LoggedInUserRequired
    @RequestMapping(method = RequestMethod.POST)
    public ResponseEntity<CreateTokenResponse> createNewToken(RequestDetails requestDetails,
                                                              @RequestParam(value = "validFor", required = false) Integer validFor) {

        ZonedDateTime now = ZonedDateTime.now(ZoneOffset.UTC);
        ZonedDateTime expiresAt = validFor == null ? now.plusDays(60) : now.plusDays(validFor);
        GeneratedTokenDetails token = tokenManagementApi.createToken(requestDetails.getRequestPermission().getCromUser(), expiresAt);

        CreateTokenResponse created = new CreateTokenResponse(
            token.getId(),
            token.getCreatedAt(),
            token.getExpiresAt(),
            token.getValue());

        return new ResponseEntity<>(created, HttpStatus.OK);
    }

    @LoggedInUserRequired
    @RequestMapping(method = RequestMethod.GET)
    public ResponseEntity<GetTokensResponse> findAllTokens(RequestDetails requestDetails) {
        List<GetTokensResponse.TokenEntryModel> tokens = tokenManagementApi
            .getTokens(requestDetails.getRequestPermission().getCromUser())
            .stream()
            .map(it -> new GetTokensResponse.TokenEntryModel(it.getId(), it.getCreatedAt(), it.getExpiresAt()))
            .collect(Collectors.toList());

        return new ResponseEntity<>(new GetTokensResponse(tokens), HttpStatus.OK);
    }
}

File: web-service/api-project/src/main/kotlin/tech/crom/service/api/user/UserTokenEndpoint.kt
package tech.crom.service.api.user

import io.ehdev.conrad.model.permission.CreateTokenResponse
import io.ehdev.conrad.model.permission.GetTokensResponse
import io.ehdev.conrad.service.api.aop.annotation.LoggedInUserRequired
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RequestParam
import tech.crom.business.api.TokenManagementApi
import tech.crom.model.token.TokenType
import tech.crom.web.api.model.RequestDetails
import java.time.ZoneOffset
import java.time.ZonedDateTime


@Controller
@RequestMapping("/api/v1/user/tokens")
open class UserTokenEndpoint @Autowired constructor(
    val tokenManagementApi: TokenManagementApi
) {

    @LoggedInUserRequired
    @RequestMapping(value = "/{tokenId}", method = arrayOf(RequestMethod.DELETE))
    open fun deleteToken(requestDetails: RequestDetails, @PathVariable("tokenId") tokenId: String): ResponseEntity<*> {
        tokenManagementApi.invalidateToken(tokenId.toLong(), TokenType.USER)
        return ResponseEntity<Any>(HttpStatus.OK)
    }

    @LoggedInUserRequired
    @RequestMapping(method = arrayOf(RequestMethod.POST))
    open fun createNewToken(requestDetails: RequestDetails,
                            @RequestParam(value = "validFor", required = false) validFor: Int?): ResponseEntity<CreateTokenResponse> {

        val now = ZonedDateTime.now(ZoneOffset.UTC)
        val expiresAt = if (validFor == null) now.plusDays(60) else now.plusDays(validFor.toLong())
        val token = tokenManagementApi.createToken(requestDetails.requestPermission.cromUser!!, expiresAt)

        val created = CreateTokenResponse(
            token.id,
            token.createdAt,
            token.expiresAt,
            token.value)

        return ResponseEntity(created, HttpStatus.CREATED)
    }

    @LoggedInUserRequired
    @RequestMapping(method = arrayOf(RequestMethod.GET))
    open fun findAllTokens(requestDetails: RequestDetails): ResponseEntity<GetTokensResponse> {
        val tokens = tokenManagementApi.getTokens(requestDetails.requestPermission.cromUser!!)
            .map({ it -> GetTokensResponse.TokenEntryModel(it.id, it.createdAt, it.expiresAt) })
            .toList()

        return ResponseEntity(GetTokensResponse(tokens), HttpStatus.OK)
    }
}
